### PR TITLE
add hideToast prop to authenticator

### PR DESCRIPTION
**Pull request by Luke-Davies**
_Issue #, if available:_
None

_Description of changes:_
Adds a property to the `amplify-authenticator` called ~showToasts~`hideToast` which is defaulted to `false`.

Setting this to `true` will prevent `amplify-toast` from rendering when `TOAST_AUTH_ERROR_EVENT` events are raised on the `UI_AUTH_CHANNEL`.

This is useful if you want to roll your own alerting components for auth error messages.

React Example:

```
import React, { useState, useEffect } from 'react';
import { Hub, HubCallback } from '@aws-amplify/core';
import {
  AmplifyAuthenticator,
} from '@aws-amplify/ui-react';
import {
 UI_AUTH_CHANNEL, TOAST_AUTH_ERROR_EVENT
} from '@aws-amplify/ui-components';

const MyAuth: React.FC = ({ children }) => {

  const [alertMessage, setAlertMessage] = useState('');

  const handleToastErrors: HubCallback = ({ payload }) => {
    if (payload.event === TOAST_AUTH_ERROR_EVENT && payload.message) {
      setAlertMessage(payload.message);
    }
  };

  useEffect(() => {
    Hub.listen(UI_AUTH_CHANNEL, handleToastErrors);
    return () => Hub.remove(UI_AUTH_CHANNEL, handleToastErrors);
  });

  return (
    <>
      {alertMessage && (<div>{alertMessage}</div>)}
      <AmplifyAuthenticator hideToast>
        // ...........
      </AmplifyAuthenticator>
    </>
  );
}
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by ThrowingSpoon** [comment]
Could this get reviewed please? Would be awesome to have 💪

**Comment by Luke-Davies** [comment]
Not sure if I'm supposed to randomly bug people for reviews sorry! @elorzafe @sammartinez @manueliglesias 

This is a pretty simple change.

Could argue that it should be `hideToasts` instead of `showToasts` to match [this prop in amplify-sign-in](https://github.com/aws-amplify/amplify-js/blob/main/packages/amplify-ui-components/src/components/amplify-sign-in/amplify-sign-in.tsx#L67). Opinions welcome.

File: packages/amplify-ui-components/src/components/amplify-authenticator/amplify-authenticator.tsx [hunk 1]
```diff
@@ -47,6 +47,8 @@ export class AmplifyAuthenticator {
   @Prop() usernameAlias: UsernameAliasStrings;
   /** Callback for Authenticator state machine changes */
   @Prop() handleAuthStateChange: AuthStateHandler = () => {};
+  /** Show amplify-toast for auth errors */
+  @Prop() showToasts: boolean = true;
```
Comment (wlee221) [inline packages/amplify-ui-components/src/components/amplify-authenticator/amplify-authenticator.tsx:51]: As you mentioned, I like `hideToasts = false` better because users would only use this prop when they want to hide them. And you can do `<amplify-authenticator hide-toasts /> too!

File: packages/amplify-ui-components/src/components/amplify-authenticator/amplify-authenticator.tsx [hunk 2]
```diff
@@ -207,7 +209,7 @@ export class AmplifyAuthenticator {
   render() {
     return (
       <Host>
-        {this.toastMessage ? (
+        {this.showToasts && this.toastMessage ? (
```
Comment (wlee221) [inline packages/amplify-ui-components/src/components/amplify-authenticator/amplify-authenticator.tsx:212]: Instead of mixing `&&` and `?`, can we use `&&` for both?
```ts
this.showToasts && this.toastMessage && (
  // ...
)
```

**Review by wlee221 — CHANGES_REQUESTED** [review]
Sorry about the late review! Just got some nits but looks good to me otherwise 👍

File: packages/amplify-ui-components/src/components/amplify-authenticator/amplify-authenticator.tsx (hunk 1 above)
Comment (wlee221) [reply packages/amplify-ui-components/src/components/amplify-authenticator/amplify-authenticator.tsx:51]: On second thought, I think we should use `hideToast` because there is only one toast at a time.

**Comment by Luke-Davies** [comment]
Thanks @wlee221! 
Both changes made and tested locally.
I've updated the PR name & description to reflect the prop change from `showToasts` --> `hideToast`

**Comment by ericclemmons** [comment]
Thanks for taking this need on @Luke-Davies.

I support @wlee221's feedback, and wanted to offer some context into past discussions around this need.

The UI components are, as we know, powered by Web Components and often rely on "slots" for customizing individual components.

What we _didn't know_ at the time is if `hideToast` would be equivalent to overriding an `amplify-toast` slot (if it existed):

```js
<AmplifyAuthenticator>
  <!-- Default components here -->
  <div slot="ampify-toast">
    <!-- Nothing to see here! -->
  </div>
</AmplifyAuthenticator>
```

Do either of y'all know if this is better, worse, or possible?  IIRC, the rationale for `hideSignUp` was because the sign-up language was "too deep" in the component to replace with an empty `slot`:
https://github.com/aws-amplify/amplify-js/pull/6098/files

Again, I'm supportive of this PR to unblock customers! I just wanted to share _where we left off_ in conversations in the past.

**Comment by Luke-Davies** [comment]
Thanks for the context @ericclemmons. I had considered trying to implement this feature by introducing a slot and would absolutely support that solution if it is more in-keeping with the design choices made for the project so far.

If a slot is introduced then I would suggest something other than `amplify-toast` for the name.

For example; our app displays a modal for auth errors because we found that some customers were not seeing the toast on certain devices unless they happened to scroll to the top of the page. Currently, we're showing this modal as well as the toast (until we can hide the toast).

Perhaps `amplify-auth-alert` or just `amplify-alert` would be a better name for the slot if that approach is preferred?

**Comment by ericclemmons** [comment]
I agree, `amplify-toast` isn't great. It matches the original desire of a "toast" notification, but we also considered `amplify-notification` among others.

I just found https://github.com/aws-amplify/amplify-js/issues/6479 as well, so we're not the only ones who've talked about this :D

So, in terms of preference having _fewer props_ would be preferential. Especially since a `slot` would allow _customization_ compared to hiding entirely.

@wlee221 What are your thoughts on the matter? Do you prefer props vs. slots, and what name would you use? (Assuming it would work!)

**Comment by wlee221** [comment]
I can't believe I missed the slot approach! I just tested that using slots works as intended, so this is really up to what we want to do.

I'm kind of divided on slots vs props though. If it were just for hiding toast, I would have argued that `hideToast` is more intuitive for customers. I find it unintuitive to pass an empty element to slot to hide a section:
```
<amplify-authenticator>
  <div slot="amplify-toast"></div>
</amplify-authenticator>
```
And it's not clear what the code is trying to do on first glance. I think the purpose of hiding an element deviates from customization, which slot specializes in. 

That being said, customers are asking for customizations on amplify-toast as well on #6479. Slots do well for covering both use cases, so I think slot is the best way to go. And I agree with @ericclemmons' comment that fewer props are preferential. 

So tldr; we should go with slots. For the slot name, I like `amplify-alert` or just `alert` because it conveys that there's an error better than other suggestions imo.

**Comment by Luke-Davies** [comment]
@wlee221 @ericclemmons 
I've raised a new PR for the slotted approach. Please see PR #7601 

I believe we can close one.

**Comment by Luke-Davies** [comment]
Coming back to this following a discussion on #7601 (`slot` approach). To understand why we've come back to `props`, see [this thread](https://github.com/aws-amplify/amplify-js/pull/7601#discussion_r588767972).

**Comment by Luke-Davies** [comment]
@wlee221 One final change I wasn't sure of on this was whether to wrap `Hub.listen(UI_AUTH_CHANNEL, this.handleToastEvent);` and `Hub.remove(UI_AUTH_CHANNEL, this.handleToastEvent);` with `if (!this.hideToast)`. The `prop` could technically be updated though (I'm not sure why anyone would) which could result in the callback not being removed.

Let me know if you think it's worth doing.

**Comment by wlee221** [comment]
I think this is a pretty simple optimization worth doing. As mentioned, I don't see a good reason for changing the `hideToast` value dynamically.

**Review by wlee221 — APPROVED** [review]
This LGTM 👍  Thanks a lot for your contribution throughout this!
